### PR TITLE
Change in overall format and explanations

### PR DESCRIPTION
**Pull request by aofarrel**
Removed the function changes from the other PR and attempted to bring this in align with Brian's notebook format rather than Terra's format. As noted on Slack, due to "reasons" this is technically untested, but as this PR is focused on form rather than function I'd rather get it out there now.

### CODE REVIEW

**Comment by aofarrel** [comment]
Now passes make and uploads successfully if requirements.sh is run within a Python 3.7 venv.

File: .gitignore [hunk 1]
```diff
@@ -6,3 +6,5 @@ __pycache__
 .idea/
 .coverage*
 .mypy_cache
+notebooks/.DS_Store
+.DS_Store
```
Comment (xbrianh) [inline .gitignore:10]: Please move .gitignore modifications to a separate PR.

File: notebooks/byod.py [hunk 2]
```diff
@@ -32,14 +36,33 @@
     | --------- | ---------  | --------- |
     | NWD1      | NWD1.cram  | NWD1.crai |
     | NWD2      | NWD2.cram  | NWD2.crai |
+
+    ## Assumptions
+    * You are not trying to overwrite a data table that already exists
```
Comment (xbrianh) [inline notebooks/byod.py:41]: Overwriting data tables rows should work fine.

File: notebooks/byod.py [hunk 3]
```diff
@@ -32,14 +36,33 @@
...
+    * Your files follow a naming convention either like this...
+    NWD119844.CRAM
+    NWD119844.CRAM.CRAI
+    ...or this:
+    NWD119844.CRAM
+    NWD119844.CRAI
+
+    Files that lack the extension .cram or .crai will not be added to the data table.
```
Comment (xbrianh) [inline notebooks/byod.py:49]: nit: I feel that this is implied.

File: notebooks/byod.py [hunk 4]
```diff
@@ -153,24 +128,14 @@ def create_cram_crai_table(table: str, listing: Iterable[str]):
...
-    """
+    # Upload to your bucket with gsutil
 
-with callysto.Cell("markdown"):
-    """
-    ## Organize your upload to Terra
-    Using instructions from the document above, you will use the gsutil tool to upload data from your local computer to
-    your Terra workspace. To effectively use the example here, we suggest you upload the data with the suggestions
+    ## Install gsutil as directed by this [document](https://support.terra.bio/hc/en-us/articles/360024056512-Uploading-to-a-workspace-Google-bucket#h_01EGP8GR3G10SKRXAC7H1ENXQ3).
+    Use the option "Set up gsutil on your local computer (step-by-step install)" which will allow you to upload files from your computer directly to Terra. Using instructions from this document, you will use the gsutil tool to upload data from your local computer to your Terra workspace. To effectively use the example here, we suggest you upload the data with the suggestions
```
Comment (xbrianh) [inline notebooks/byod.py:134]: This paragraph seems to contain redundant information. Can it be reworded?

File: notebooks/byod.py [hunk 5]
```diff
@@ -183,16 +148,16 @@ def create_cram_crai_table(table: str, listing: Iterable[str]):
...
-    `gsutil cp /Users/Documents/Example.cram gs://your_bucket_info/test-crai-cram/`
-    """
+    `gsutil cp /Users/my-cool-username/Documents/Example.cram gs://your_bucket_info/my-crams/`
+
+    We will be calling what comes after your bucket info, here represented as `my-crams`, as your sudirectory.
 
-with callysto.Cell("markdown"):
-    """
     ## Preview the data in your workspace bucket
+    Be aware that if you have uploaded multiple files, all of them will appear with this ls command. It will also contain one folder for every workflow you have run in this workspace. You may want to skip this step if you're after uploading hundreds of files. However, if you have imported data tables from Gen3, they will not show up here as the files within are only downloaded when their associated TSV tables are called upon by workflows or iPython notebooks.
```
Comment (xbrianh) [inline notebooks/byod.py:156]: Users should be encouraged to explore the contents of their workspace bucket, not discouraged. Also, large output is scrollable in Jupyter cells. Even long listings will not overwhelm the screen.

Please consider removing information extraneous to this notebook, such as discussion about workflow folders and Gen3 tables.

File: notebooks/byod.py [hunk 6]
```diff
@@ -214,13 +179,34 @@ def create_cram_crai_table(table: str, listing: Iterable[str]):
...
+    | NWD3             | gs://my-workspace-bucket/my-crams/NWD3.cram          | gs://my-workspace-bucket/my-crams/NWD3.crai |
+    """
+
+with callysto.Cell("python"):
+    listing = [key for key in gs.list_bucket("my-crams")]
+    create_cram_crai_table("my-table-name", listing)
+
+with callysto.Cell("markdown"):
+    """
+    Now, go check the data section of your workspace and you should a data table with the name you have given it, and that table should act as a directory of your files.
```
Comment (xbrianh) [inline notebooks/byod.py:205]: grammar.

File: notebooks/byod.py [hunk 7]
```diff
@@ -214,13 +179,34 @@ def create_cram_crai_table(table: str, listing: Iterable[str]):
...
+
+    If you set the name of your table to a table that already exists, the old table will not be overwritten, but the new table won't be created either. Terra tables cannot be updated, they must be deleted and remade.
```
Comment (xbrianh) [inline notebooks/byod.py:207]: I don't think this is true. You can check this yourself using the methods `upload_data_table`, `upload_rows`, or `upload_columns` defined in this notebook.

File: notebooks/byod.py [hunk 8]
```diff
@@ -214,13 +179,34 @@ def create_cram_crai_table(table: str, listing: Iterable[str]):
...
+    Note: If you get a FireCloudServerError that reads "Duplicated entities are not allowed in TSV", this may be because you ran the code more than once without deleting `final.tsv` first. If you don't delete that file before re-running the notebook, Firecloud will consider your second upload to be a duplicate and block it.
```
Comment (xbrianh) [inline notebooks/byod.py:208]: This sentence appears unrelated to the current byod material.

File: requirements.sh [hunk 9]
```diff
@@ -0,0 +1,6 @@
+# Only run inside a Python 3.7 venv!
+pip install flake8
+pip install mypy
+pip install firecloud
+pip install git+https://github.com/xbrianh/callysto
+pip install terra-notebook-utils
```
Comment (xbrianh) [inline requirements.sh:6]: Please remove: requirements are now handled differently.

**Review by xbrianh — CHANGES_REQUESTED** [review]
Looks like good progress, but there are a few things to address.

Also, please consider wrapping lines longer than 120 characters. This will make it easier to read content on GitHub.

File: .gitignore (hunk 1 above)
Comment (aofarrel) [reply .gitignore:10]: This was mostly here for my own sanity. Unfortunately I either have to add them to .gitignore or the files will show up in the PR because they change every time you so much as breathe on a Mac. I've noticed that neither you nor Lon have them in .gitignore though -- how do you get around .DS_Store showing up everywhere?

File: notebooks/byod.py (hunk 7 above)
Comment (aofarrel) [reply notebooks/byod.py:207]: Interesting... it definitely did work like this when I wrote my own functions, so I assumed it was the same for yours as both rely on Firecloud. I'll double check with some tests like you suggested, but I wouldn't be surprised if this was updated to match the behavior you're expecting because it was definitely frustrating to have to rename the output notebook every time you ran it. Thanks for catching this.

File: notebooks/byod.py (hunk 3 above)
Comment (aofarrel) [reply notebooks/byod.py:49]: I'm of two minds about this one... @bethsheets, your thoughts? It may be implicit, but a researcher might also be unsure if txt files (common if someone has a lot of data) will "mess up" the table.

File: notebooks/byod.py (hunk 5 above)
Comment (aofarrel) [reply notebooks/byod.py:156]: I find the scrolling on Terra notebooks to be a bit unresponsive (at least for me -- I zoom almost all webpages in by 10-20%, which can break internal scrolls sometimes), so I shy away from it. But you make a good point -- let's have them explore. I'll reword this as less of a warning and more of an explanation.

File: .flake8 [hunk 10]
```diff
@@ -1,3 +1,3 @@
 [flake8]
 max-line-length=120
-ignore: E301, E302, E305, E401, F401, W503, W504, W605, E252, E402, E501, E265
+ignore: E301, E302, E305, E401, F401, W503, W504, W605, E252, E402, E501, E265, W291
```
Comment (DailyDreaming) [inline .flake8:3]: I don't think we should commit the sin of trailing white spaces.  >.>

Comment (aofarrel) [reply .flake8:3]: But I LIKE trailing whitespace. 😈 

This is a duplicate of https://github.com/DataBiosphere/bdcat_notebooks/pull/27#discussion_r496213696, but Brian pointed out it might not be needed on Slack. Basically, the issue is that markdown is not standardized, so *sometimes* markdown requires a double space for a newline but not always. But it looks like  <br/> works in Jupyter so that's good enough for me!

File: notebooks/byod.py [hunk 11]
```diff
@@ -153,53 +125,76 @@ def create_cram_crai_table(table: str, listing: Iterable[str]):
...
-    Using instructions from the document above, you will use the gsutil tool to upload data from your local computer to
-    your Terra workspace. To effectively use the example here, we suggest you upload the data with the suggestions
-    below.
-    """
+    ## Install gsutil as directed by this [document](https://support.terra.bio/hc/en-us/articles/360024056512-Uploading-to-a-workspace-Google-bucket#h_01EGP8GR3G10SKRXAC7H1ENXQ3).
+    Use the option "Set up gsutil on your local computer (step-by-step install)" which will allow you to upload files 
+    from your computer directly to Terra. (Those familiar with Terra are aware that files can also be uploaded by
+    simply dragging them in to the "Files" section in the Data tab, but as `gsutil` can upload more than one file at
+    a time and has additional error handling, it is recommended to use it over dragging.) But before you can upload, 
+    we will need to perform a few quick tasks.
```
Comment (bethsheets) [inline notebooks/byod.py:136]: For the phrase "Those familiar with Terra are aware"-- I think we should avoid language assuming what a user should be familiar with as it can be invalidating. I suggest something like: "If you're uploading only a few files, you can use Terra's drag and drop feature; for many files we suggest using `gsutil` ."

File: notebooks/byod.py [hunk 12]
```diff
@@ -153,53 +125,76 @@ def create_cram_crai_table(table: str, listing: Iterable[str]):
...
     """
+    #### A note on subdirectories
+    Subdirectories are not true directories. If all files in a subdirectory are deleted, it will cease to exist. That is
+    to say, Google Cloud does not have any equivalent to empty folders. If you would like to know more, [Google has
+    documentation on its filesystem's inner workings](https://cloud.google.com/storage/docs/gsutil/addlhelp/HowSubdirectoriesWork), 
+    but most users will not need to know the details.
+
+    ## Begin the upload
+    Now that you know what you will name your subdirectory, turn back to your computer's terminal and upload the data
+    using `gsutil cp`. Remember you may have to log into to your Google account before using gsutil into your
```
Comment (bethsheets) [inline notebooks/byod.py:167]: The phrasing "before using gsutil into your workspace bucket"  is difficult to follow. Please rephrase.

File: notebooks/byod.py [hunk 13]
```diff
@@ -153,53 +125,76 @@ def create_cram_crai_table(table: str, listing: Iterable[str]):
...
+    We will be calling what comes after your bucket info, here represented as `my-crams`, as your sudirectory. We'll be
+    using that subdirectory name later on, so let's make note of it here.
+
+    """
+with callysto.Cell("python"):
+    subdirectory = "my-crams"
 with callysto.Cell("markdown"):
     """
+    #### A note on subdirectories
+    Subdirectories are not true directories. If all files in a subdirectory are deleted, it will cease to exist. That is
```
Comment (bethsheets) [inline notebooks/byod.py:160]: Give context that this "cloud subdirectories". 

Is subdirectories the correct/common terminology? I'm not sure.

File: notebooks/byod.py [hunk 14]
```diff
@@ -153,53 +125,76 @@ def create_cram_crai_table(table: str, listing: Iterable[str]):
...
+    but most users will not need to know the details.
+
+    ## Begin the upload
+    Now that you know what you will name your subdirectory, turn back to your computer's terminal and upload the data
+    using `gsutil cp`. Remember you may have to log into to your Google account before using gsutil into your
+    workspace bucket -- you can do so with `gcloud auth` as described in [Google's documentation](https://cloud.google.com/sdk/gcloud/reference/auth/login).
+
     ## Preview the data in your workspace bucket
+    Let's first look at the top of the workspace bucket. This will match what you see if you go the "data" tab of
+    your Terra workspace and click "Files" under the heading "OTHER DATA." There will be one directory per WDL workflow
```
Comment (bethsheets) [inline notebooks/byod.py:172]: Are we expecting users to have already executed a WDL workflow? I would remove -- it is tangential and confusing for first-time users trying to get their data set up. You could rephrase to clarify "all data you have uploaded or generated in your Terra workspace bucket will be listed in the Files section."

File: notebooks/byod.py [hunk 15]
```diff
@@ -153,53 +125,76 @@ def create_cram_crai_table(table: str, listing: Iterable[str]):
...
+    that has executed. You will also see your subdirectory. However, if you have imported data 
+    tables from Gen3, they will not show up here, as the files within are only downloaded when their associated
```
Comment (bethsheets) [inline notebooks/byod.py:174]: Again, explaining a Gen3 example here is tangential to the main message and users may not know what you're talking about. I would remove.

File: notebooks/byod.py [hunk 16]
```diff
@@ -259,9 +272,69 @@ def create_cram_crai_table(table: str, listing: Iterable[str]):
...
     Note that the row for `NWD2` is missing from the combined table since it was not present in `diabetic_table`.
+    """
 
+with callysto.Cell("markdown"):
+    """
+    # Additional functions
+    To aid in the creation of your own data tables, we have provided some more functions for you to use and adapt.
     """
 
+with callysto.Cell("python"):
```
Comment (bethsheets) [inline notebooks/byod.py:283]: Could we ask Brian for brief descriptions of each of these functions that can be included as comments?

**Review by bethsheets — CHANGES_REQUESTED** [review]
This is looking really nice! I added some comments to the instructions to clarify and avoid bringing up other complex topics that we don't introduce here. 

Instead of opening another PR, I also wanted to mention a typo in 5.1 "In addition to seeing the to being able to see the datatable in Terra directly."

File: notebooks/byod.py (hunk 13 above)
Comment (aofarrel) [reply notebooks/byod.py:160]: There doesn't appear to be any correct terminology, or even consistent terminology. That's why my older documentation called it psuedofolders. But you're right, I should make this clearer that it is specific to the cloud.

File: notebooks/byod.py (hunk 16 above)
Comment (xbrianh) [reply notebooks/byod.py:293]: Yes! We can handle this in a separate PR.

**Review by xbrianh — APPROVED** [review]
